### ISSUE TITLE
add ability to disable implicitly prepared queries.

### ISSUE THREAD

**Issue by james-lawrence** [jackc/pgx#331]
it looks like pgx implicitly prepares all queries particularly within the stdlib interface, which is both surprising and can sometimes be undesirable. I understand that pgx gets its performance boost from using them implicitly.

Would we be open to making this behaviour configurable?

Details:
ran into this while debugging pgbouncer queries returning a `ERROR: unsupported pkt type: 80 (SQLSTATE 08P01)` when querying pgbouncer's internal state. I'm assuming the type is the ascii code for message format's first byte, which would make it a [parse message](https://www.postgresql.org/docs/current/static/protocol-message-formats.html).
```golang
func (c *Conn) Query(query string, argsV []driver.Value) (driver.Rows, error) {
	if !c.conn.IsAlive() {
		return nil, driver.ErrBadConn
	}

	ps, err := c.conn.Prepare("", query)
	if err != nil {
		return nil, err
	}

	restrictBinaryToDatabaseSqlTypes(ps)

	return c.queryPrepared("", argsV)
}
```

**Comment by jackc** [comment]
I think that is a reasonable option.

There is already support for using the simple protocol / not using prepared statements via the `QueryEx` method. To do this by default we would need an additional option in `ConnConfig` -- something like `DisableImplicitPrepare`. Then `QueryEx` and `execEx` would need to to consider that flag in addition to the `options.SimpleProtocol`.

To use this new functionality, the stdlib connection would need to be opened with that configuration option. That can be done with `stdlib.RegisterDriverConfig`.

**Comment by james-lawrence** [comment]
k I'll take a look and see about generating a PR at somepoint.

**Comment by james-lawrence** [comment]
I started looking into this, not sure how to handle the call to restrict binary types as it is attached to a prepared statement. any suggestions?

**Comment by james-lawrence** [comment]
this is the current query function for stdlib:
```golang
func (c *Conn) Query(query string, argsV []driver.Value) (driver.Rows, error) {
	if !c.conn.IsAlive() {
		return nil, driver.ErrBadConn
	}

	ps, err := c.conn.Prepare("", query)
	if err != nil {
		return nil, err
	}

	restrictBinaryToDatabaseSqlTypes(ps)

	return c.queryPrepared("", argsV)
}```

see the restrictBinaryToDatabaseSqlTypes? it takes a prepared statement, the same statement I'm trying to add a way of not performing. I don't see a way of replicating the behaviour of the restrict function,looking for suggestions from you on how you'd like me to procede.

**Comment by james-lawrence** [comment]
If I'm reading ConnInfo correctly and how its used, then changes to it wouldn't be necessary. It'd just need to be properly configured during connection initialization for the stdlib interface to replicate the restrictBinaryToDatabaseSqlTypes.

**Comment by jackc** [comment]
Ahh. I was thinking strictly in terms of the native interface.

Hmmm. This is interesting. The problem is approaching it with restricting binary types is even if all types are using the text protocol it still would use the prepared statement interface.

stdlib will need to branch and not use prepared statements. It might be slightly messy wiring all this together... Not sure the best approach.

**Comment by james-lawrence** [comment]
TBH, I don't mind just exploring and seeing what I come up with. Just wanted to check in if I was missing something obvious. =)

**Comment by james-lawrence** [comment]
found further issues, simple query also seems to create a prepared statement under the hood.

```
func (c *Conn) sendSimpleQuery(sql string, args ...interface{}) error {
	if err := c.ensureConnectionReadyForQuery(); err != nil {
		return err
	}

	if len(args) == 0 {
		buf := appendQuery(c.wbuf, sql)

		_, err := c.conn.Write(buf)
		if err != nil {
			c.die(err)
			return err
		}
		c.pendingReadyForQueryCount++

		return nil
	}

	ps, err := c.Prepare("", sql)
	if err != nil {
		return err
	}

	return c.sendPreparedQuery(ps, args...)
}
```

**Comment by james-lawrence** [comment]
as it stand I don't believe it is possible to use pgx without generating a prepared statement.

to do so would require an entirely new function on the conn. do you agree?

**Comment by jackc** [comment]
It definitely **is** possible to send individual queries that do not use prepared statements. The question is how could it conveniently be made to do it all the time.

A single query can be sent without using a prepared statement by setting `QueryExOptions.SimpleProtocol` to `true`. When that is set, the query parameters are sanitized and interpolated into the SQL string and sent as a simple query.

**Comment by jackc** [comment]
As an addendum let me clarify my above remarks are for the native interface. The standard library interface does not expose the simple protocol at this point.

**Comment by james-lawrence** [comment]
> It definitely is possible to send individual queries that do not use prepared statements.

hmm, not what I am seeing with regards to SimpleProtocol in particular, there does seem to be a code path when the ParameterOID's are empty, but thats not terrible useful.

following the code from QueryEx:
```golang
func (c *Conn) QueryEx(ctx context.Context, sql string, options *QueryExOptions, args ...interface{}) (rows *Rows, err error) {
	// snipped ...
	if options != nil && options.SimpleProtocol {
		err = c.sanitizeAndSendSimpleQuery(sql, args...)
		if err != nil {
			rows.fatal(err)
			return rows, err
		}

		return rows, nil
	}
	// snipped ...
}
```
```golang
func (c *Conn) sanitizeAndSendSimpleQuery(sql string, args ...interface{}) (err error) {
	// snipped ...
	return c.sendSimpleQuery(sql)
}
```
```golang
func (c *Conn) sendSimpleQuery(sql string, args ...interface{}) error {
	if err := c.ensureConnectionReadyForQuery(); err != nil {
		return err
	}
	// this branch isn't useful as it only triggers when there are no arguments to the query.
	if len(args) == 0 {
		buf := appendQuery(c.wbuf, sql)

		_, err := c.conn.Write(buf)
		if err != nil {
			c.die(err)
			return err
		}
		c.pendingReadyForQueryCount++

		return nil
	}
	// and here we hit another prepared statement.
	ps, err := c.Prepare("", sql)
	if err != nil {
		return err
	}

	return c.sendPreparedQuery(ps, args...)
}
```

looking at the c.Prepare function, maybe the fact the prepared statement name is blank treats it as a normal query with performing the round trip to postgres to actually create a real prepared query?

if thats the case then this becomes very easy.

**Comment by jackc** [comment]
Take a closer look at how `sanitizeAndSendSimpleQuery` and `sendSimpleQuery` interact. `sanitizeAndSendSimpleQuery` accepts a sql string and arguments and sanitizes and interpolates the arguments into a new sql string. Then it calls `sendSimpleQuery` with that new sql string and no arguments.

`sendSimpleQuery` uses the simple protocol (no prepared statement) when it has no args.

```go
	if len(args) == 0 {
		buf := appendQuery(c.wbuf, sql)

		_, err := c.conn.Write(buf)
		if err != nil {
			c.die(err)
			return err
		}
		c.pendingReadyForQueryCount++

		return nil
	}
```

**Comment by james-lawrence** [comment]
huh, tricky =) that explains the oddity I was seeing while poking around during implementation where the list of args ended up being empty.

anyways I managed to get a working implementation. have 2 PRs, I've opened one against pgx. its just a minor refactor to stdlib restrict binary types.

the other is here https://github.com/james-lawrence/pgx/pull/1 which works but we may want to drop/change parts of it due to the hacked together process I follow while working. but it'd open up a cleaner way to create drivers that follow PG wire protocol.
